### PR TITLE
return endpoints that is in the same namespace as its referred pod

### PR DESCRIPTION
**Pull request by yangminzhu**

### CODE REVIEW

File: pilot/pkg/serviceregistry/kube/controller.go [hunk 1]
```diff
@@ -409,8 +409,13 @@ func (c *Controller) InstancesByPort(hostname model.Hostname, reqSvcPort int,
...
 // GetProxyServiceInstances returns service instances co-located with a given proxy
 func (c *Controller) GetProxyServiceInstances(proxy *model.Proxy) ([]*model.ServiceInstance, error) {
-	var out []*model.ServiceInstance
 	proxyIP := proxy.IPAddress
+	proxyNS := ""
+	if pod, exists := c.pods.getPodByIP(proxyIP); exists {
+		proxyNS = pod.Namespace
+	}
+	endpointsForPodInSameNS := make([]*model.ServiceInstance, 0)
+	endpointsForPodInDifferentNS := make([]*model.ServiceInstance, 0)
```
Comment (rshriram) [inline pilot/pkg/serviceregistry/kube/controller.go]: why do we need the second one?

File: pilot/pkg/serviceregistry/kube/controller.go [hunk 2]
```diff
@@ -423,22 +428,31 @@ func (c *Controller) GetProxyServiceInstances(proxy *model.Proxy) ([]*model.Serv
 			continue
 		}
 
+		endpoints := &endpointsForPodInSameNS
+		if ep.Namespace != proxyNS {
```
Comment (rshriram) [inline pilot/pkg/serviceregistry/kube/controller.go]: this should never occur

Comment (yangminzhu) [reply pilot/pkg/serviceregistry/kube/controller.go]: I think the naming is a little misleading, probably I should change the `proxyNS` to something like `podNS`.
But basically, I'm trying to put the [k8s endpoint](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#endpoints-v1-core) which has a different namespace compared to the referred pod to have lower priority when Pilot is generating listeners.

Also note it's possible if you define a [k8s service](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#service-v1-core) but doesn't specify labels, and then define a [k8s endpoint](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#endpoints-v1-core) that has the same name as the `k8s service`, in the `k8s endpoint` you can use IP address to bind it with pod, even if it's not in the endpoint's namespace.

File: pilot/pkg/serviceregistry/kube/controller.go (hunk 1 above)
Comment (yangminzhu) [reply pilot/pkg/serviceregistry/kube/controller.go]: Please see my reply in your other comment.

File: pilot/pkg/serviceregistry/kube/controller.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller.go]: @yangminzhu This is not a right way to use endpoint and svc. If we do that we can not access the service by cluster ip(
```
root@szvp000201060:~# iptables-save |grep  10.0.0.133 
-A KUBE-SERVICES -d 10.0.0.133/32 -p tcp -m comment --comment "default/reviews:http has no endpoints" -m tcp --dport 9080 -j REJECT --reject-with icmp-port-unreachable

root@szvp000201060:~# kubectl get ep -n istio-system reviews
NAME      ENDPOINTS                                           AGE
reviews   172.17.0.18:9080,172.17.0.19:9080,172.17.0.5:9080   2m43s
```
).
Also we can not access it in istio mesh, I have tested this with bookinfo example.

File: pilot/pkg/serviceregistry/kube/controller.go [hunk 3]
```diff
@@ -409,8 +409,13 @@ func (c *Controller) InstancesByPort(hostname model.Hostname, reqSvcPort int,
...
 func (c *Controller) GetProxyServiceInstances(proxy *model.Proxy) ([]*model.ServiceInstance, error) {
-	var out []*model.ServiceInstance
 	proxyIP := proxy.IPAddress
+	podNamespace := ""
+	if pod, exists := c.pods.getPodByIP(proxyIP); exists {
+		podNamespace = pod.Namespace
+	}
+	endpointsForPodInSameNS := make([]*model.ServiceInstance, 0)
+	endpointsForPodInDifferentNS := make([]*model.ServiceInstance, 0)
 	for _, item := range c.endpoints.informer.GetStore().List() {
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller.go]: BTW, not relate to this pr.
 I have another doubt: each proxy we do iterate on all endpoints and services, is that efficient? Can we do some optimization？

File: pilot/pkg/serviceregistry/kube/controller.go (hunk 2 above)
Comment (yangminzhu) [reply pilot/pkg/serviceregistry/kube/controller.go]: Could you be more specific what do you want for this comment?
For my previous comment, It's just a way (I'm not saying it's a right way or not) to use endpoint that is allowed by k8s, you would be still able to access the service by its name directly.

Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller.go]: I mean this `if ep.Namespace != proxyNS {` should never happen. If it does, then the svc can not be accessed.

Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller.go]: So, I think we do not need these code in this section.

**Pull request merged**